src/basic/path-util.c

```diff
@@ -639,14 +639,31 @@ static int check_x_access(const char *path, int *ret_fd) {
         return 0;
 }
 
-int find_executable_full(const char *name, bool use_path_envvar, char **ret_filename, int *ret_fd) {
+int find_executable_full(const char *name, const char *root, bool use_path_envvar, char **ret_filename, int *ret_fd) {
         int last_error, r;
         const char *p = NULL;
 
         assert(name);
 
         if (is_path(name)) {
                 _cleanup_close_ int fd = -1;
+                _cleanup_free_ char *path_name = NULL;
+
+                /* Function chase_symlinks is invoked only when root is not NULL,
+                 * as using it regardless of root value would alter the behavior
+                 * of existing callers */
+                if (root) {
+                        r = chase_symlinks(name,
+                                           root,
+                                           CHASE_PREFIX_ROOT|CHASE_SAFE|CHASE_NOFOLLOW,
+                                           &path_name,
+                                           /* ret_fd= */ NULL); /* prefix root to name in case full paths are not specified */
+                        if (r < 0)
+                                return r;
+                }
+
+                if (path_name)
+                        name = path_name;
```

Comment:
why this in a separate if block? the only way this is non-NULL is if chase_symlinks() succeeded, hence just do the assignment unconditionally right after the chase_symlinks() when it succeeded